File: packages/javascript/core/converter.js
/**
 * Chain Converter: Bidirectional CodeUChain Transformation
 *
 * Converts between CodeUChain (chain-based) and traditional imperative code.
 * Enables zero-overhead execution while maintaining maintainability.
 *
 * Features:
 * - CodeUChain → Traditional: Generates optimized imperative code with direct function calls
 * - Traditional → CodeUChain: Reconstructs chain definitions from imperative patterns
 * - Validation: Ensures both versions produce identical results
 */

const { Context, Chain, Link } = require('../core');

class ChainConverter {
  /**
   * Bidirectional converter between CodeUChain and traditional imperative code.
   *
   * @example
   * // Convert chain to traditional code
   * const converter = new ChainConverter();
   * const traditionalCode = converter.chainToCode(myChain);
   *
   * // Convert traditional code back to chain
   * const reconstructedChain = converter.codeToChain(traditionalCode);
   *
   * // Validate both produce same results
   * const isValid = await converter.validate(myChain, reconstructedChain, testContexts);
   */
  constructor() {
    this.indent = '  ';
  }

  /**
   * Convert a CodeUChain chain to optimized traditional imperative code.
   *
   * This generates zero-overhead code with direct function calls,
   * eliminating the chain orchestration layer at runtime.
   *
   * @param {Chain} chain - The CodeUChain chain to convert
   * @param {string} functionName - Name for the generated function
   * @param {boolean} includeJsDoc - Whether to include JSDoc comments
   * @param {boolean} includeClasses - Whether to include link class definitions
   * @returns {string} JavaScript source code as a string
   *
   * @example
   * const chain = new Chain();
   * chain.addLink(new ValidateLink(), 'validate');
   * chain.addLink(new ProcessLink(), 'process');
   * const converter = new ChainConverter();
   * const code = converter.chainToCode(chain);
   * console.log(code);
   */
  chainToCode(chain, functionName = 'executePipeline', includeJsDoc = true, includeClasses = false) {
    const linksInfo = this._extractLinksInfo(chain);

    const lines = [];

    // Add header comment
    lines.push('// Auto-generated from CodeUChain');
    lines.push('// Zero-overhead traditional code');
    lines.push('');

    // Optionally include link class implementations
    if (includeClasses) {
      for (const linkInfo of linksInfo) {
        if (linkInfo.sourceCode) {
          lines.push(linkInfo.sourceCode);
          lines.push('');
        }
      }
    }

    // Function signature
    if (includeJsDoc) {
      lines.push('/**');
      lines.push(' * Zero-overhead execution of the chain logic.');
      lines.push(' *');
      lines.push(' * This function provides identical functionality to the original chain');
      lines.push(' * but with direct function calls for maximum performance.');
      lines.push(' *');
      lines.push(' * @param {Object} initialData - Initial context data');
      lines.push(' * @returns {Promise<Object>} Final context data');
      lines.push(' */');
    }
    lines.push(`async function ${functionName}(initialData) {`);

    // Initialize context
    lines.push(`${this.indent}// Initialize context data`);
    lines.push(`${this.indent}let ctxData = { ...initialData };`);
    lines.push('');

    // Generate direct calls for each link
    let linkIndex = 0;
    for (const linkInfo of linksInfo) {
      linkIndex++;
      lines.push(`${this.indent}// Link ${linkIndex}: ${linkInfo.name}`);

      // Create link instance
      const className = linkInfo.linkInstance.constructor.name;
      lines.push(`${this.indent}const ${linkInfo.name}Link = new ${className}();`);

      // Create context and call
      lines.push(`${this.indent}const ${linkInfo.name}Ctx = new Context(ctxData);`);
      lines.push(`${this.indent}const ${linkInfo.name}Result = await ${linkInfo.name}Link.call(${linkInfo.name}Ctx);`);
      lines.push(`${this.indent}ctxData = ${linkInfo.name}Result.toObject();`);
      lines.push('');
    }

    // Return final context
    lines.push(`${this.indent}return ctxData;`);
    lines.push('}');

    return lines.join('\n');
  }

  /**
   * Extract information about all links in the chain.
   *
   * @private
   * @param {Chain} chain - The chain to extract from
   * @returns {Array<Object>} Array of link information objects
   */
  _extractLinksInfo(chain) {
    const linksInfo = [];

    for (const [name, link] of chain._links.entries()) {
      const linkInfo = {
        name: name,
        linkInstance: link,
        callMethod: link.call,
        sourceCode: this._extractSourceCode(link)
      };

      linksInfo.push(linkInfo);
    }

    return linksInfo;
  }

  /**
   * Extract source code for a link class.
   *
   * @private
   * @param {Link} link - The link to extract source from
   * @returns {string|null} Source code or null if unavailable
   */
  _extractSourceCode(link) {
    try {
      // Get the class source from its toString
      const className = link.constructor.name;
      const classCode = link.constructor.toString();

      // Format as a class declaration
      if (classCode.includes('class ')) {
        return classCode;
      } else {
        // It's a function constructor, convert to class syntax
        return `class ${className} {\n${classCode}\n}`;
      }
    } catch (error) {
      return null;
    }
  }

  /**
   * Convert traditional imperative code back to a CodeUChain chain.
   *
   * This analyzes the code structure and reconstructs the chain definition,
   * enabling maintenance and modification using the CodeUChain framework.
   *
   * Note: This is a simplified implementation that extracts link patterns
   * from code comments. For full AST-based parsing, consider using a
   * dedicated parser library.
   *
   * @param {string} code - JavaScript source code to analyze
   * @param {string} functionName - Name of the function to convert
   * @returns {Chain} A reconstructed CodeUChain chain
   *
   * @example
   * const code = `
   * async function executePipeline(initialData) {
   *   let ctxData = { ...initialData };
   *   // Link 1: validate
   *   const validateLink = new ValidateLink();
   *   const validateCtx = new Context(ctxData);
   *   const validateResult = await validateLink.call(validateCtx);
   *   ctxData = validateResult.toObject();
   *   return ctxData;
   * }
   * `;
   * const converter = new ChainConverter();
   * const chain = converter.codeToChain(code);
   */
  codeToChain(code, functionName = 'executePipeline') {
    // Extract link patterns using regex (simplified approach)
    const linkPattern = /const\s+(\w+)Link\s*=\s*new\s+(\w+)\(\)/g;
    const links = [];
    
    let match;
    while ((match = linkPattern.exec(code)) !== null) {
      const linkName = match[1];
      const className = match[2];
      links.push({ linkName, className });
    }

    // Build the chain
    const chain = new Chain();

    // Try to execute the code to get the link classes
    try {
      // Use Function constructor to evaluate code
      const evalFunc = new Function('Context', 'Chain', 'Link', `
        ${code}
        return { ${links.map(l => l.className).join(', ')} };
      `);

      const linkClasses = evalFunc(Context, Chain, Link);

      // Add links to chain
      for (const linkInfo of links) {
        if (linkClasses[linkInfo.className]) {
          const linkInstance = new linkClasses[linkInfo.className]();
          chain.addLink(linkInstance, linkInfo.linkName);
        }
      }
    } catch (error) {
      // If we can't evaluate, return empty chain structure
      console.warn(`Could not evaluate code: ${error.message}`);
    }

    return chain;
  }

  /**
   * Validate that chain and converted function produce identical results.
   *
   * @param {Chain} chain - Original CodeUChain chain
   * @param {Function} convertedFunc - Converted traditional function
   * @param {Array<Object>} testContexts - List of test input contexts
   * @param {number} tolerance - Tolerance for floating-point comparisons
   * @returns {Promise<Object>} Object with {isValid, errors}
   *
   * @example
   * const chain = myChain;
   * const code = converter.chainToCode(chain);
   * eval(code);
   * const testData = [{ x: 1 }, { x: 2 }, { x: 3 }];
   * const result = await converter.validate(chain, executePipeline, testData);
   * console.log(result.isValid, result.errors);
   */
  async validate(chain, convertedFunc, testContexts, tolerance = 1e-9) {
    const errors = [];

    for (let i = 0; i < testContexts.length; i++) {
      const testCtx = testContexts[i];

      try {
        // Run chain
        const chainResult = await chain.run(new Context(testCtx));
        const chainData = chainResult.toObject();

        // Run converted function
        const funcResult = await convertedFunc(testCtx);

        // Compare results
        if (!this._compareObjects(chainData, funcResult, tolerance)) {
          errors.push(
            `Test case ${i}: Results differ\n` +
            `  Chain result: ${JSON.stringify(chainData)}\n` +
            `  Function result: ${JSON.stringify(funcResult)}`
          );
        }
      } catch (error) {
        errors.push(`Test case ${i}: Exception occurred: ${error.message}`);
      }
    }

    return {
      isValid: errors.length === 0,
      errors: errors
    };
  }

  /**
   * Compare two objects with tolerance for floating-point values.
   *
   * @private
   * @param {Object} obj1 - First object to compare
   * @param {Object} obj2 - Second object to compare
   * @param {number} tolerance - Tolerance for floating-point comparisons
   * @returns {boolean} True if objects are equal within tolerance
   */
  _compareObjects(obj1, obj2, tolerance) {
    if (obj1 === obj2) return true;
    if (obj1 == null || obj2 == null) return false;
    if (typeof obj1 !== typeof obj2) return false;

    if (typeof obj1 === 'number' && typeof obj2 === 'number') {
      return Math.abs(obj1 - obj2) <= tolerance;
    }

    if (Array.isArray(obj1) && Array.isArray(obj2)) {
      if (obj1.length !== obj2.length) return false;
      for (let i = 0; i < obj1.length; i++) {
        if (!this._compareObjects(obj1[i], obj2[i], tolerance)) {
          return false;
        }
      }
      return true;
    }

    if (typeof obj1 === 'object' && typeof obj2 === 'object') {
      const keys1 = Object.keys(obj1);
      const keys2 = Object.keys(obj2);

      if (keys1.length !== keys2.length) return false;

      for (const key of keys1) {
        if (!keys2.includes(key)) return false;
        if (!this._compareObjects(obj1[key], obj2[key], tolerance)) {
          return false;
        }
      }

      return true;
    }

    return obj1 === obj2;
  }

  /**
   * Generate an optimized class that encapsulates the chain logic.
   *
   * This creates a reusable class with zero-overhead execution,
   * suitable for production use.
   *
   * @param {Chain} chain - The CodeUChain chain to convert
   * @param {string} className - Name for the generated class
   * @param {boolean} includeClasses - Whether to include link class definitions
   * @returns {string} JavaScript source code for the optimized class
   *
   * @example
   * const chain = myChain;
   * const code = converter.generateOptimizedClass(chain, 'MyOptimizedPipeline');
   * eval(code);
   * const pipeline = new MyOptimizedPipeline();
   * const result = await pipeline.execute({ input: 'data' });
   */
  generateOptimizedClass(chain, className = 'OptimizedPipeline', includeClasses = false) {
    const linksInfo = this._extractLinksInfo(chain);

    const lines = [];

    // Add header
    lines.push('// Auto-generated optimized pipeline class');
    lines.push('');

    // Optionally include link class definitions
    if (includeClasses) {
      for (const linkInfo of linksInfo) {
        if (linkInfo.sourceCode) {
          lines.push(linkInfo.sourceCode);
          lines.push('');
        }
      }
    }

    // Class definition
    lines.push(`class ${className} {`);
    lines.push(`${this.indent}/**`);
    lines.push(`${this.indent} * Zero-overhead optimized pipeline.`);
    lines.push(`${this.indent} *`);
    lines.push(`${this.indent} * This class provides the same functionality as the original chain`);
    lines.push(`${this.indent} * but with direct method calls for maximum performance.`);
    lines.push(`${this.indent} */`);
    lines.push('');

    // Constructor
    lines.push(`${this.indent}constructor() {`);
    lines.push(`${this.indent}${this.indent}// Initialize link instances`);
    for (const linkInfo of linksInfo) {
      const linkClassName = linkInfo.linkInstance.constructor.name;
      lines.push(`${this.indent}${this.indent}this.${linkInfo.name}Link = new ${linkClassName}();`);
    }
    lines.push(`${this.indent}}`);
    lines.push('');

    // Execute method
    lines.push(`${this.indent}/**`);
    lines.push(`${this.indent} * Execute the pipeline with zero overhead.`);
    lines.push(`${this.indent} * @param {Object} initialData - Initial context data`);
    lines.push(`${this.indent} * @returns {Promise<Object>} Final context data`);
    lines.push(`${this.indent} */`);
    lines.push(`${this.indent}async execute(initialData) {`);
    lines.push(`${this.indent}${this.indent}let ctxData = { ...initialData };`);
    lines.push('');

    let linkIndex = 0;
    for (const linkInfo of linksInfo) {
      linkIndex++;
      lines.push(`${this.indent}${this.indent}// Link ${linkIndex}: ${linkInfo.name}`);
      lines.push(`${this.indent}${this.indent}const ${linkInfo.name}Ctx = new Context(ctxData);`);
      lines.push(`${this.indent}${this.indent}const ${linkInfo.name}Result = await this.${linkInfo.name}Link.call(${linkInfo.name}Ctx);`);
      lines.push(`${this.indent}${this.indent}ctxData = ${linkInfo.name}Result.toObject();`);
      lines.push('');
    }

    lines.push(`${this.indent}${this.indent}return ctxData;`);
    lines.push(`${this.indent}}`);
    lines.push('}');

    return lines.join('\n');
  }
}

module.exports = { ChainConverter };

File: packages/python/codeuchain/utils/converter.py
"""
Chain Converter: Bidirectional CodeUChain Transformation

Converts between CodeUChain (chain-based) and traditional imperative code.
Enables zero-overhead execution while maintaining maintainability.

Features:
- CodeUChain → Traditional: Generates optimized imperative code with direct function calls
- Traditional → CodeUChain: Reconstructs chain definitions from imperative patterns
- Validation: Ensures both versions produce identical results
"""

import ast
import inspect
import textwrap
from typing import Any, Dict, List, Optional, Callable, Type
from dataclasses import dataclass

from ..core import Context, Chain, Link


@dataclass
class LinkInfo:
    """Information about a link in the chain."""
    name: str
    link_instance: Any
    call_method: Callable
    source_code: Optional[str] = None


class ChainConverter:
    """
    Bidirectional converter between CodeUChain and traditional imperative code.
    
    Usage:
        # Convert chain to traditional code
        converter = ChainConverter()
        traditional_code = converter.chain_to_code(my_chain)
        
        # Convert traditional code back to chain
        reconstructed_chain = converter.code_to_chain(traditional_code)
        
        # Validate both produce same results
        is_valid = converter.validate(my_chain, reconstructed_chain, test_contexts)
    """
    
    def __init__(self):
        self._indent = "    "
    
    def chain_to_code(
        self, 
        chain: Chain, 
        function_name: str = "execute_pipeline",
        include_types: bool = True
    ) -> str:
        """
        Convert a CodeUChain chain to optimized traditional imperative code.
        
        This generates zero-overhead code with direct function calls,
        eliminating the chain orchestration layer at runtime.
        
        Args:
            chain: The CodeUChain chain to convert
            function_name: Name for the generated function
            include_types: Whether to include type hints
            
        Returns:
            Python source code as a string
            
        Example:
            >>> chain = Chain()
            >>> chain.add_link(ValidateLink(), "validate")
            >>> chain.add_link(ProcessLink(), "process")
            >>> converter = ChainConverter()
            >>> code = converter.chain_to_code(chain)
            >>> print(code)
        """
        links_info = self._extract_links_info(chain)
        
        # Build the function
        lines = []
        
        # Add imports
        lines.append("# Auto-generated from CodeUChain")
        lines.append("# Zero-overhead traditional code")
        lines.append("from typing import Any, Dict")
        lines.append("")
        
        # Extract and add link implementations
        for link_info in links_info:
            if link_info.source_code:
                lines.append(link_info.source_code)
                lines.append("")
        
        # Function signature
        if include_types:
            lines.append(f"async def {function_name}(initial_data: Dict[str, Any]) -> Dict[str, Any]:")
        else:
            lines.append(f"async def {function_name}(initial_data):")
        
        lines.append(f'{self._indent}"""')
        lines.append(f"{self._indent}Zero-overhead execution of the chain logic.")
        lines.append(f"{self._indent}")
        lines.append(f"{self._indent}This function provides identical functionality to the original chain")
        lines.append(f"{self._indent}but with direct function calls for maximum performance.")
        lines.append(f'{self._indent}"""')
        
        # Initialize context
        lines.append(f"{self._indent}# Initialize context data")
        lines.append(f"{self._indent}ctx_data = initial_data.copy()")
        lines.append("")
        
        # Generate direct calls for each link
        for i, link_info in enumerate(links_info):
            lines.append(f"{self._indent}# Link {i+1}: {link_info.name}")
            
            # Create link instance
            class_name = link_info.link_instance.__class__.__name__
            lines.append(f"{self._indent}{link_info.name}_link = {class_name}()")
            
            # Create context and call
            lines.append(f"{self._indent}{link_info.name}_ctx = Context(ctx_data)")
            lines.append(f"{self._indent}{link_info.name}_result = await {link_info.name}_link.call({link_info.name}_ctx)")
            lines.append(f"{self._indent}ctx_data = {link_info.name}_result.to_dict()")
            lines.append("")
        
        # Return final context
        lines.append(f"{self._indent}return ctx_data")
        
        return "\n".join(lines)
    
    def _extract_links_info(self, chain: Chain) -> List[LinkInfo]:
        """Extract information about all links in the chain."""
        links_info = []
        
        for name, link in chain._links.items():
            # Get the call method
            call_method = getattr(link, 'call', None)
            
            # Try to get source code
            source_code = None
            try:
                source = inspect.getsource(link.__class__)
                source_code = textwrap.dedent(source)
            except (OSError, TypeError):
                # Can't get source for built-in or C-extension classes
                pass
            
            links_info.append(LinkInfo(
                name=name,
                link_instance=link,
                call_method=call_method,
                source_code=source_code
            ))
        
        return links_info
    
    def code_to_chain(
        self, 
        code: str,
        function_name: str = "execute_pipeline"
    ) -> Chain:
        """
        Convert traditional imperative code back to a CodeUChain chain.
        
        This analyzes the code structure and reconstructs the chain definition,
        enabling maintenance and modification using the CodeUChain framework.
        
        Args:
            code: Python source code to analyze
            function_name: Name of the function to convert
            
        Returns:
            A reconstructed CodeUChain chain
            
        Example:
            >>> code = '''
            ... async def execute_pipeline(initial_data):
            ...     ctx_data = initial_data.copy()
            ...     # Link 1: validate
            ...     validate_link = ValidateLink()
            ...     validate_ctx = Context(ctx_data)
            ...     validate_result = await validate_link.call(validate_ctx)
            ...     ctx_data = validate_result.to_dict()
            ...     return ctx_data
            ... '''
            >>> converter = ChainConverter()
            >>> chain = converter.code_to_chain(code)
        """
        # Parse the code
        tree = ast.parse(code)
        
        # Find the target function
        func_node = None
        for node in ast.walk(tree):
            if isinstance(node, ast.AsyncFunctionDef) and node.name == function_name:
                func_node = node
                break
        
        if not func_node:
            raise ValueError(f"Function '{function_name}' not found in code")
        
        # Extract link patterns from the function body
        links = self._extract_links_from_ast(func_node)
        
        # Build the chain
        chain = Chain()
        
        # Execute the code to get the link classes
        namespace = {}
        exec(code, namespace)
        
        # Add links to chain
        for link_name, link_class_name in links:
            if link_class_name in namespace:
                link_instance = namespace[link_class_name]()
                chain.add_link(link_instance, link_name)
        
        return chain
    
    def _extract_links_from_ast(self, func_node: ast.AsyncFunctionDef) -> List[tuple]:
        """Extract link information from the AST of a function."""
        links = []
        
        for node in ast.walk(func_node):
            # Look for patterns like: link_name_link = ClassName()
            if isinstance(node, ast.Assign):
                if len(node.targets) == 1:
                    target = node.targets[0]
                    if isinstance(target, ast.Name) and target.id.endswith('_link'):
                        if isinstance(node.value, ast.Call):
                            if isinstance(node.value.func, ast.Name):
                                link_name = target.id.replace('_link', '')
                                class_name = node.value.func.id
                                links.append((link_name, class_name))
        
        return links
    
    async def validate(
        self, 
        chain: Chain, 
        converted_func: Callable,
        test_contexts: List[Dict[str, Any]],
        tolerance: float = 1e-9
    ) -> tuple[bool, List[str]]:
        """
        Validate that chain and converted function produce identical results.
        
        Args:
            chain: Original CodeUChain chain
            converted_func: Converted traditional function
            test_contexts: List of test input contexts
            tolerance: Tolerance for floating-point comparisons
            
        Returns:
            Tuple of (is_valid, error_messages)
            
        Example:
            >>> chain = my_chain
            >>> code = converter.chain_to_code(chain)
            >>> exec(code, globals())
            >>> test_data = [{"x": 1}, {"x": 2}, {"x": 3}]
            >>> is_valid, errors = await converter.validate(chain, execute_pipeline, test_data)
        """
        errors = []
        
        for i, test_ctx in enumerate(test_contexts):
            try:
                # Run chain
                chain_result = await chain.run(Context(test_ctx))
                chain_data = chain_result.to_dict()
                
                # Run converted function
                func_result = await converted_func(test_ctx)
                
                # Compare results
                if not self._compare_dicts(chain_data, func_result, tolerance):
                    errors.append(
                        f"Test case {i}: Results differ\n"
                        f"  Chain result: {chain_data}\n"
                        f"  Function result: {func_result}"
                    )
            except Exception as e:
                errors.append(f"Test case {i}: Exception occurred: {str(e)}")
        
        return len(errors) == 0, errors
    
    def _compare_dicts(
        self, 
        dict1: Dict[str, Any], 
        dict2: Dict[str, Any],
        tolerance: float
    ) -> bool:
        """Compare two dictionaries with tolerance for floating-point values."""
        if set(dict1.keys()) != set(dict2.keys()):
            return False
        
        for key in dict1.keys():
            val1, val2 = dict1[key], dict2[key]
            
            # Handle floating-point comparison
            if isinstance(val1, (float, int)) and isinstance(val2, (float, int)):
                if abs(float(val1) - float(val2)) > tolerance:
                    return False
            # Handle nested dicts
            elif isinstance(val1, dict) and isinstance(val2, dict):
                if not self._compare_dicts(val1, val2, tolerance):
                    return False
            # Handle lists
            elif isinstance(val1, list) and isinstance(val2, list):
                if len(val1) != len(val2):
                    return False
                for v1, v2 in zip(val1, val2):
                    if isinstance(v1, dict) and isinstance(v2, dict):
                        if not self._compare_dicts(v1, v2, tolerance):
                            return False
                    elif v1 != v2:
                        return False
            # Direct comparison
            elif val1 != val2:
                return False
        
        return True
    
    def generate_optimized_class(
        self,
        chain: Chain,
        class_name: str = "OptimizedPipeline"
    ) -> str:
        """
        Generate an optimized class that encapsulates the chain logic.
        
        This creates a reusable class with zero-overhead execution,
        suitable for production use.
        
        Args:
            chain: The CodeUChain chain to convert
            class_name: Name for the generated class
            
        Returns:
            Python source code for the optimized class
            
        Example:
            >>> chain = my_chain
            >>> code = converter.generate_optimized_class(chain, "MyOptimizedPipeline")
            >>> exec(code)
            >>> pipeline = MyOptimizedPipeline()
            >>> result = await pipeline.execute({"input": "data"})
        """
        links_info = self._extract_links_info(chain)
        
        lines = []
        
        # Add imports
        lines.append("# Auto-generated optimized pipeline class")
        lines.append("from typing import Any, Dict")
        lines.append("")
        
        # Extract link class definitions
        for link_info in links_info:
            if link_info.source_code:
                lines.append(link_info.source_code)
                lines.append("")
        
        # Class definition
        lines.append(f"class {class_name}:")
        lines.append(f'{self._indent}"""')
        lines.append(f"{self._indent}Zero-overhead optimized pipeline.")
        lines.append(f"{self._indent}")
        lines.append(f"{self._indent}This class provides the same functionality as the original chain")
        lines.append(f"{self._indent}but with direct method calls for maximum performance.")
        lines.append(f'{self._indent}"""')
        lines.append("")
        
        # Constructor
        lines.append(f"{self._indent}def __init__(self):")
        lines.append(f"{self._indent}{self._indent}# Initialize link instances")
        for link_info in links_info:
            class_name_str = link_info.link_instance.__class__.__name__
            lines.append(f"{self._indent}{self._indent}self.{link_info.name}_link = {class_name_str}()")
        lines.append("")
        
        # Execute method
        lines.append(f"{self._indent}async def execute(self, initial_data: Dict[str, Any]) -> Dict[str, Any]:")
        lines.append(f'{self._indent}{self._indent}"""Execute the pipeline with zero overhead."""')
        lines.append(f"{self._indent}{self._indent}ctx_data = initial_data.copy()")
        lines.append("")
        
        for i, link_info in enumerate(links_info):
            lines.append(f"{self._indent}{self._indent}# Link {i+1}: {link_info.name}")
            lines.append(f"{self._indent}{self._indent}{link_info.name}_ctx = Context(ctx_data)")
            lines.append(f"{self._indent}{self._indent}{link_info.name}_result = await self.{link_info.name}_link.call({link_info.name}_ctx)")
            lines.append(f"{self._indent}{self._indent}ctx_data = {link_info.name}_result.to_dict()")
            lines.append("")
        
        lines.append(f"{self._indent}{self._indent}return ctx_data")
        
        return "\n".join(lines)
